### DISCUSSION TITLE
Create a solid along a path with IfcSurfaceCurveSweptAreaSolid

### DISCUSSION THREAD

**Discussion by arunkvijay** [IfcOpenShell/IfcOpenShell#2951] [Q&A] [answered]
Hi,
How do I create a swept solid with `IfcSurfaceCurveSweptAreaSolid` and `IfcArbitraryClosedProfileDef` either using BlenderBim or IfcOpenshell.
I have a profile defined with `IfcArbitraryClosedProfileDef` with profile type as `AREA`
Is it possible to create a sweptsolid with this profile and the swept path defined with another `IfcArbitraryClosedProfileDef` with profile type as `CURVE`

**Comment by aothms** [comment]
>  Is it possible to create a sweptsolid with this profile and the swept path defined with another IfcArbitraryClosedProfileDef with profile type as CURVE

No, the attribute type of IfcSurfaceCurveSweptAreaSolid.Directrix is IfcCurve. IfcArbitraryClosedProfileDef[ProfileType=CURVE] is not an IfcCurve. The example in the documentation can probably be a IfcTrimmedCurve of IfcEllipse with a IfcCylindericalSurface as the ReferenceSurface. Does that help, what kind of help do you need precisely?

**Reply by arunkvijay** [reply]
Hi Thomas, I am trying to create members which is swept along a path, something like this post [https://forums.buildingsmart.org/t/ifcsurfacecurvesweptareasolid-does-not-consistently-orient-the-swept-profile/1533](https://forums.buildingsmart.org/t/ifcsurfacecurvesweptareasolid-does-not-consistently-orient-the-swept-profile/1533) 
I have created member type with `IfcArbitraryClosedProfileDef` that I would like to sweep along a closed path.
Any pointers on how to do this through ifcopenshell.

**Reply by aothms** [reply]
Do you have a code snippet that illustrates where you are stuck?

**Reply by arunkvijay** [reply]
Hi Thomas, I am still learning and exploring ifcopenshell. I have created an `IfcMember` on blenderbim with `IfcArbitraryClosedProfileDef` and I want to sweep this member along a path to create as indicated in the attachment below. For this I was looking at ifcopenshell python api and wanted help in pointing me to the right apis that I could use.
Could you help me on this.

![Screenshot from 2023-04-11 06-32-52](https://user-images.githubusercontent.com/58535690/231029963-cc686a21-f6e3-4c2f-82bb-b7c771459be9.png)
![Screenshot from 2023-04-11 06-12-32](https://user-images.githubusercontent.com/58535690/231029976-005f508d-d141-4ee6-84d2-df07606ba018.png)

**Reply by aothms** [reply, marked as answer]
Ok, thanks for the context. I don't think we have a high level API for that kind of representation, so it boils down to manipulating the instances directly. I wanted to work on an example and illustration for the spec anyway so here is the code:

```python
import ifcopenshell
import ifcopenshell.template

from math import *

f = ifcopenshell.template.create()

scsas = f.createIfcSurfaceCurveSweptAreaSolid(
    SweptArea=f.createIfcRectangleProfileDef(
        "AREA",
        None,
        None,
        0.1,
        0.1,
    ),
    Position=None,
    Directrix=f.createIfcTrimmedCurve(
        f.createIfcEllipse(
            f.createIfcAxis2Placement3D(
                f.createIfcCartesianPoint((0.0, 0.0, 0.0)),
                f.createIfcDirection((1.0, 0.0, 1.0)),
                f.createIfcDirection((1.0, 0.0, 0.0)),
            ),
            sqrt(2.0) / 2.0,
            0.5,
        ),
        [f.createIfcParameterValue(180.)],
        [f.createIfcParameterValue(0.0)],
        True,
        'PARAMETER'
    ),
    ReferenceSurface=f.createIfcCylindricalSurface(
        f.createIfcAxis2Placement3D(f.createIfcCartesianPoint((0.0, 0.0, 0.0))),
        0.5,
    ),
)

project = f.by_type("IfcProject")[0]
oh = f.by_type("IfcOwnerHistory")[0]
context = f.by_type("IfcRepresentationContext")[0]

site = f.createIfcSite(
    ifcopenshell.guid.new(),
    oh,
    ObjectPlacement=f.createIfcLocalPlacement(
        None, f.createIfcAxis2Placement3D(f.createIfcCartesianPoint((0.0, 0.0, 0.0)))
    ),
)

building = f.createIfcBuilding(
    ifcopenshell.guid.new(),
    oh,
    ObjectPlacement=f.createIfcLocalPlacement(
        site.ObjectPlacement,
        f.createIfcAxis2Placement3D(f.createIfcCartesianPoint((0.0, 0.0, 0.0))),
    ),
)

beam = f.createIfcBeam(
    ifcopenshell.guid.new(),
    oh,
    ObjectPlacement=f.createIfcLocalPlacement(
        building.ObjectPlacement,
        f.createIfcAxis2Placement3D(f.createIfcCartesianPoint((0.0, 0.0, 0.0))),
    ),
    Representation=f.createIfcProductDefinitionShape(
        None,
        None,
        [
            f.createIfcShapeRepresentation(
                context, "Body", "AdvancedSweptSolid", Items=[scsas]
            )
        ],
    ),
)

f.createIfcRelContainedInSpatialStructure(
    ifcopenshell.guid.new(),
    oh,
    None,
    None,
    [beam],
    building,
)

f.createIfcRelAggregates(ifcopenshell.guid.new(), oh, None, None, site, [building])
f.createIfcRelAggregates(ifcopenshell.guid.new(), oh, None, None, project, [site])

f.write("surface-curve-swept-area.ifc")
```

Looks like this:

![afbeelding](https://user-images.githubusercontent.com/1096535/231096545-1eeb2167-0d51-4870-b7be-34ea0a534786.png)

Don't expect miracles in off the shelves tools. These kind of advanced representations aren't always supported well. (A somewhat old probably version of BimVision on my machine crashed on this file).

**Reply by arunkvijay** [reply]
Thanks a lot for the code @aothms . This is very helpful.

One more question, the ifcopenshell api documentation shows entity creation as  `ifcopenshell.api.run("root.create_entity", f, ifc_class='IfcSite')`, whereas you have mentioned entity creation as `f.createIfcSite`. Are these both the same. The relationship aggregation is also done in your code as `f.createIfcRelAggregates`. I feel your way of code is much more direct. But, I don't see any documentation for these codes in `ifcopenshell` api. Where do I get to read/refer about all these. Any pointers would be helpful
